client/beefy/src/communication/request_response/outgoing_requests_engine.rs

```diff
@@ -0,0 +1,245 @@
+		while let Some(peer) = self.peers_cache.pop_front() {
+			if live.contains(&peer) {
+				return Some(peer)
+			}
+		}
+		None
+	}
+
+	fn request_from_peer(&mut self, peer: PeerId, block: NumberFor<B>) {
+		debug!(target: "beefy::sync", "🥩 requesting justif #{:?} from peer {:?}", block, peer);
+
+		let payload = JustificationRequest::<B> { begin: block }.encode();
+
+		let (tx, rx) = oneshot::channel();
+
+		self.network.start_request(
+			peer,
+			self.protocol_name.clone(),
+			payload,
+			tx,
+			IfDisconnected::ImmediateError,
+		);
+
+		self.state = State::AwaitingResponse(peer, block, stream::once(rx));
+	}
+
+	/// If no other request is in progress, start new justification request for `block`.
+	pub fn request(&mut self, block: NumberFor<B>) {
+		// ignore new requests while there's already one pending
+		match &self.state {
+			State::AwaitingResponse(_, _, _) => return,
+			State::Idle(_) => (),
+		}
+		self.reset_peers_cache_for_block(block);
+
+		// Start the requests engine - each unsuccessful received response will automatically
+		// trigger a new request to the next peer in the `peers_cache` until there are none left.
+		if let Some(peer) = self.try_next_peer() {
+			self.request_from_peer(peer, block);
+		} else {
+			debug!(target: "beefy::sync", "🥩 no good peers to request justif #{:?} from", block);
+		}
+	}
+
+	/// Cancel any pending request for block numbers smaller or equal to `block`.
+	pub fn cancel_requests_older_than(&mut self, block: NumberFor<B>) {
+		match &self.state {
+			State::AwaitingResponse(_, number, _) if *number <= block => {
+				debug!(
+					target: "beefy::sync",
+					"🥩 cancel pending request for justification #{:?}",
+					number
+				);
+				self.state = State::Idle(stream::pending());
+			},
+			_ => (),
+		}
+	}
+
+	fn process_response(
+		&mut self,
+		peer: PeerId,
+		block: NumberFor<B>,
+		validator_set: &ValidatorSet<AuthorityId>,
+		response: Result<Response, Canceled>,
+	) -> Result<BeefyVersionedFinalityProof<B>, Error> {
+		response
+			.map_err(|e| {
+				debug!(
+					target: "beefy::sync",
+					"🥩 for on demand justification #{:?}, peer {:?} hung up: {:?}",
+					block, peer, e
+				);
+				Error::InvalidResponse
+			})?
+			.map_err(|e| {
+				debug!(
+					target: "beefy::sync",
+					"🥩 for on demand justification #{:?}, peer {:?} error: {:?}",
+					block, peer, e
+				);
+				Error::InvalidResponse
+			})
+			.and_then(|encoded| {
+				decode_and_verify_finality_proof::<B>(&encoded[..], block, &validator_set).map_err(
+					|e| {
+						debug!(
+							target: "beefy::sync",
+							"🥩 for on demand justification #{:?}, peer {:?} responded with invalid proof: {:?}",
+							block, peer, e
+						);
+						Error::InvalidResponse
+					},
+				)
+			})
+	}
+
+	pub async fn next(&mut self) -> Option<BeefyVersionedFinalityProof<B>> {
+		let (peer, block, resp) = match &mut self.state {
+			State::Idle(pending) => {
+				let _ = pending.next().await;
```

Comment:
`futures::pending!()`

Is what you want here. Then you could also replace `State` with an `Option`.